rgw: add new cap user-info-without-keys

alimaredia opened the pull request:
This new cap removes S3 keys and Swift keys from
the response in the admin api command 'get user info'.





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

thotz commented on src/rgw/driver/rados/rgw_rest_user.cc:
```diff
@@ -111,7 +113,15 @@ void RGWOp_User_Info::execute(optional_yield y)
-  op_ret = RGWUserAdminOp_User::info(s, driver, op_state, flusher, y);
+  // dump_keys is false if users-without-keys is 'read' and
+  // the user is not the system user
+  int keys_perm = s->user->get_info().caps.check_cap("users-without-keys", RGW_CAP_READ);
```
Any specific reason to not returning the error in case of RGW_CAP_WRITE?

alimaredia replied:
The intention is that there's no point in setting the cap to `write` because it wouldn't allow any operations from happening that were already not allowed. Setting the cap to `write` just means it's not set to `read`, and the `read` bit is the only bit we care about.

Making a config variable variable that just hid the keys crossed my mind, but I needed something that could be applied to a particular user, and would be uniform across all RGWs. A new cap made the most sense.

cbodley commented on doc/radosgw/adminops.rst:
```diff
@@ -273,13 +273,15 @@ TBD.
 =============
 
-Get user information.
+Get user information. Cap ``users`` must be set to ``read`` to run this operation.
+If additional cap ``users-without-keys`` is set to ``read`` or ``*``, S3 Key and Swift keys will not be
+included in the response unless the user running this operation is the system user.
```
i didn't expect `users-without-keys` to *remove* permissions this way. no other caps work like this

i thought the admin was going to give out `users-without-keys` caps *instead* of `users` when they shouldn't have access to keys

so `RGWOp_User_Info::check_caps()` would grant access if the user had either `users` or `users-without-keys` read caps, but omit keys for the latter

alimaredia replied:
That was the other way I was thinking of approaching it. The only conflict that would occur is if both `users` and `users-without-keys` were set (on accident) and the read bit was set on each too.

In that case do you agree that the effect of the `users-without-keys` cap would prevail on a `get user info` call?

cbodley replied:
permissions should be additive, so if `users=read` grants read access to everything, then `users=read`+`users-without-keys=read` should too

cbodley commented on src/rgw/driver/rados/rgw_rest_user.cc:
```diff
@@ -111,7 +121,15 @@ void RGWOp_User_Info::execute(optional_yield y)
+  // dump_keys is false if users-without-keys is 'read' and
+  // the user is not the system user
+  int keys_perm = s->user->get_info().caps.check_cap("users-without-keys", RGW_CAP_READ);
+  if (keys_perm == 0 && !op_state.system) {
```
in addition to system users, admin users should also see the keys by default. you can test for admin access with something like `s->auth.identity->is_admin_of(uid)`

alimaredia replied:
What is an admin user? A user with the cap `users` set that created a user? I can't find any upstream documentation on that.

cbodley commented on src/rgw/driver/rados/rgw_rest_user.cc:
```diff
@@ -635,6 +677,10 @@ class RGWOp_Key_Create : public RGWRESTOp {
   RGWOp_Key_Create() {}
 
   int check_caps(const RGWUserCaps& caps) override {
+    int r = caps.check_cap("users-without-keys", RGW_CAP_WRITE);
```
"users-without-keys" shouldn't grant access to RGWOp_Key_Create or RGWOp_Key_Remove

cbodley commented on src/rgw/driver/rados/rgw_rest_user.cc:
```diff
@@ -738,6 +788,10 @@ class RGWOp_Caps_Add : public RGWRESTOp {
   RGWOp_Caps_Add() {}
 
   int check_caps(const RGWUserCaps& caps) override {
+    int r = caps.check_cap("users-without-keys", RGW_CAP_WRITE);
```
a user with `users-without-keys=write` could use this RGWOp_Caps_Add to grant themselves `users=*` caps and circumvent the 'without-keys' restriction. let's require "users" for this one?

i think a similar argument applies for RGWOp_User_Create, which could create a new user with `users=*` caps (or the system flag) then use that to view/edit keys

cbodley commented on src/rgw/driver/rados/rgw_rest_user.cc:
```diff
@@ -111,7 +121,15 @@ void RGWOp_User_Info::execute(optional_yield y)
-  op_ret = RGWUserAdminOp_User::info(s, driver, op_state, flusher, y);
+  // dump_keys is false if users-without-keys is 'read' and
+  // the user is not the system user
+  int keys_perm = s->user->get_info().caps.check_cap("users", RGW_CAP_READ);
+  if (keys_perm == 0 || op_state.system || s->auth.identity->is_admin_of(uid)) {
+    dump_keys = true;
```
👍

alimaredia commented:
I squashed all of the commits and a run in teuthology passed all of the tests I added: https://pulpito.ceph.com/amaredia-2024-01-31_22:07:55-rgw:singleton-wip-rgw-add-cap-user-without-keys-distro-default-smithi/

cbodley commented on src/rgw/driver/rados/rgw_user.cc:
```diff
@@ -125,8 +125,8 @@ static void dump_swift_keys_info(Formatter *f, RGWUserInfo &info)
 
-static void dump_user_info(Formatter *f, RGWUserInfo &info,
-                           RGWStorageStats *stats = NULL)
+static void dump_user_info(const DoutPrefixProvider* dpp, Formatter *f, RGWUserInfo &info,
+                           bool dump_keys, RGWStorageStats *stats = NULL)
```
the `dpp` argument is unused?

alimaredia commented:
@cbodley new run with the `dpp` removed: https://pulpito.ceph.com/amaredia-2024-02-05_21:50:20-rgw:singleton-wip-rgw-add-cap-user-without-keys-distro-default-smithi/

alimaredia commented:
jenkins test api

alimaredia commented:
Run of the rgw suite: https://pulpito.ceph.com/amaredia-2024-02-08_15:23:19-rgw-wip-rgw-add-cap-user-without-keys-distro-default-smithi/
